lib/util.c

```diff
@@ -2193,3 +2193,137 @@ EXPORTED char *modseqtoa(modseq_t modseq)
     buf_printf(&buf, MODSEQ_FMT, modseq);
     return buf_release(&buf);
 }
+
+static char *_xsyslog_ev_escape(const char *val)
+{
+    struct buf buf = BUF_INITIALIZER;
+    int needs_escaping = 0, orig_len, escaped_len;
+    const char *p;
+
+    buf_setcstr(&buf, val);
+
+    escaped_len = orig_len = buf_len(&buf);
+
+    // Make sure the empty string is visible
+    if (orig_len == 0) {
+        buf_setcstr(&buf, "\"\"");
+        return buf_release(&buf);
+    }
+
+    for (p = buf_cstring(&buf); *p; p++) {
+        switch (*p) {
+        case '\\':
+        case '\"':
+        case '\n':
+        case '\r':
+            ++escaped_len;  // add 1 for the backslash
+
+        // FALL THROUGH
+        case ' ':
+            needs_escaping = 1;
+            break;
+        }
+    }
+
+    if (needs_escaping) {
+        char *q;
+
+        escaped_len += 2;  // add 2 for surrounding DQUOTEs
+
+        buf_truncate(&buf, escaped_len);  // grow the buffer to escaped length
+
+        // we can now build the escaped value in place, tail to head
+        q = (char *) buf_base(&buf) + escaped_len - 1;
+        *q-- = '\"';  // closing DQUOTE
+
+        for (p = buf_base(&buf) + orig_len - 1; p >= buf_base(&buf); p--) {
+            char c = *p;
+
+            switch (c) {
+            case '\\':
+            case '\"':
+                needs_escaping = 1;
+                break;
+
+            case '\n':
+                needs_escaping = 1;
+                c = 'n';
+                break;
+
+            case '\r':
+                needs_escaping = 1;
+                c = 'r';
+                break;
+
+            default:
+                needs_escaping = 0;
+                break;
+            }
+
+            *q-- = c;
+
+            if (needs_escaping) *q-- = '\\';
+        }
+
+        assert(q == buf_base(&buf));
+        *q = '\"';  // opening DQUOTE
+    }
+
+    return buf_release(&buf);
+}
+
+EXPORTED void _xsyslog_ev(int priority, const char *event,
+                          logfmt_arg_list *arg)
+{
+    static struct buf buf = BUF_INITIALIZER;
+    int saved_errno = errno;
+
+    char *escaped = _xsyslog_ev_escape(event);
+    buf_printf(&buf, "event=%s", escaped);
```

Comment:
It's a micro-optimization in light of the heap allocations later in this function and maybe it's even better to keep `buf_print` for consistency with the LF_ formatting below, but: in general we tend to avoid `buf_printf` for simple formatting. It's surprisingly costly when compared to

```
buf_setcstr(&buf, "event=");
buf_appendcstr(&buf, escaped);
```

Reply:
I've made some changes here, see what you think?